src/java.base/unix/native/criuengine/criuengine.c

```diff
@@ -61,6 +61,75 @@ static int kickjvm(pid_t jvm, int code) {
     return 0;
 }
 
+static char *join_args(const char **args) {
+    size_t size = 0;
+    for (const char **argp = args; *argp; ++argp) {
+      size += strlen(*argp) * 4 + 2 + 1;
+    }
+    char *retval = malloc(size);
+    if (!retval) {
+        perror("malloc");
+        exit(1);
+    }
+    char *d = retval;
+    for (const char **argp = args; *argp; ++argp) {
+        if (d != retval) {
+            *d++ = ' ';
+        }
+        // https://unix.stackexchange.com/a/357932/296319
+        if (!strpbrk(*argp, " \t\n!\"#$&'()*,;<=>?[\\]^`{|}~")) {
+            d = stpcpy(d, *argp);
+            continue;
+        }
+        *d++ = '\'';
+        for (const char *s = *argp; *s; ++s) {
+          if (*s != '\'') {
+            *d++ = *s;
+          } else {
+            d = stpcpy(d, "'\\''");
+          }
+        }
+        *d++ = '\'';
+    }
+    *d = 0;
+    return retval;
+}
+
+static char *path_join(const char *path1, const char *path2) {
+    char *retval;
+    if (path2[0] == '/') {
+        retval = strdup(path2);
+        if (!retval) {
+            perror("strdup");
+            exit(1);
+        }
+        return retval;
+    }
+    if (asprintf(&retval, "%s/%s", path1, path2) == -1) {
+        perror("asprintf");
+        exit(1);
+    }
+    return retval;
+}
+
+static char *path_abs(const char *rel) {
+    char *cwd = get_current_dir_name();
+    if (!cwd) {
+        perror("get_current_dir_name");
+        exit(1);
+    }
+    char *retval = path_join(cwd, rel);
+    free(cwd);
+    return retval;
+}
+
+static char *path_abs2(const char *rel1, const char *rel2) {
+    char *rel1_abs = path_abs(rel1);
+    char *retval = path_join(rel1_abs, rel2);
+    free(rel1_abs);
+    return retval;
+}
```

Comment:
This function is closer to path_from, rather than to abs. But by using just path_join/path_from, we'll provide a shorter yet still correct path to the log (the only use of the function), so I propose to drop it.

Reply:
It became now simple enough in C++, is it OK?